src/nupic/algorithms/SpatialPoolerExtended.cpp

```diff
@@ -0,0 +1,108 @@
+
+#include <nupic/algorithms/SpatialPoolerExtended.hpp>
+using namespace nupic::algorithms::spatial_pooler_extended;
+
+SpatialPoolerExtended::SpatialPoolerExtended(
+        const vector<UInt>  inputDimensions,
+        const vector<UInt>  columnDimensions,
+        UInt                potentialRadius,
+        Real                potentialPct,
+        bool                wrapAround,
+        Real                localAreaDensity,
+        UInt                stimulusThreshold,
+        Real                synPermInactiveDec,
+        Real                synPermActiveInc,
+        Real                synPermConnected,
+        UInt                dutyCyclePeriod,
+        Real                minPctOverlapDutyCycles,
+        Int                 seed,
+        UInt                spVerbosity)
+    : SpatialPooler(
+        /* inputDimensions */               inputDimensions,
+        /* columnDimensions */              columnDimensions,
+        /* potentialRadius */               potentialRadius,
+        /* potentialPct */                  potentialPct,
+        /* globalInhibition */              true,
+        /* localAreaDensity */              localAreaDensity,
+        /* numActiveColumnsPerInhArea */    -1,
+        /* stimulusThreshold */             stimulusThreshold,
+        /* synPermInactiveDec */            synPermInactiveDec,
+        /* synPermActiveInc */              synPermActiveInc,
+        /* synPermConnected */              synPermConnected,
+        /* minPctOverlapDutyCycles */       minPctOverlapDutyCycles,
+        /* dutyCyclePeriod */               dutyCyclePeriod,
+        /* boostStrength */                 0.0f,
+        /* seed */                          seed,
+        /* spVerbosity */                   spVerbosity,
+        /* wrapAround */                    wrapAround)
+{
+    activeDutyCycles_.assign(numColumns_, localAreaDensity_);
+}
+
+// TODO: initialize
+
+// TODO: Rework potential pools
+
+void SpatialPoolerExtended::boostOverlaps_(const vector<UInt> &overlaps, //TODO use Eigen sparse vector here
+                                   vector<Real> &boosted) const {
+  const Real denominator = 1.0f / log2( localAreaDensity_ );
+  for (UInt i = 0; i < numColumns_; i++) {
+    // Apply tiebreakers before boosting, so that boosting is applied to the
+    // tiebreakers.  This is important so that the tiebreakers don't hurt the
+    // entropy of the result by biasing some mini-columns to activte more often
+    // than others.
+    boosted[i] = (overlaps[i] + tieBreaker_[i])
+                 * log2(activeDutyCycles_[i]) * denominator;
+  }
+}
+
+void SpatialPoolerExtended::inhibitColumnsGlobal_(const vector<Real> &overlaps,
+                          Real density,
+                          vector<UInt> &activeColumns) const
+{
+  NTA_ASSERT(!overlaps.empty());
+  NTA_ASSERT(density > 0.0f && density <= 1.0f);
+  const UInt miniColumns = columnDimensions_.back();
```

Comment:
now I probably understand, SP colDims {100, 100, 40} does not mean a 3D SP, but that out of the 100x100 columns, we take 40 mini-columns in each macro col. Right? 
If so, please do not piggyback on columnDimensions like that. Add a new param `UInt numMiniColumnsPerMacroColumn`. If you want to keep the API the same between original SP and this one, set default =1 and add it there.

Reply:
Yes, and it's a hack.  This piggybacking is why I didn't need to rewrite the topology (which I really should have done by now) but it just so happens that with the MNIST dataset this works.

I agree that numMiniColumns should be split into its own parameter, though the result of the SP will still be a single large SDR with the mini-columns as the final dimensions (as it is now).